src/debug/dwarf/line_numbers.cr

```diff
@@ -389,20 +389,54 @@ module Debug
         end
       end
 
+      # When decoding statement programs when asking for the current
+      # filename it's usually the case that it's the same as the old
+      # one (because all info from a single file comes first, then
+      # another file comes next, etc.). So we remember the last
+      # mapped index to avoid unnecessary lookups.
+      @last_filename : String?
+      @last_filename_index = 0
+
       private def register_filename(name)
-        if index = @files.index(name)
-          return index
+        if name.same?(@last_filename)
+          return @last_filename_index
         end
-        @files << name
-        @files.size - 1
+
+        @last_filename = name
+
+        index = @files.index(name)
```

Comment:
How often does it go back and look up a file that it has already seen? If it happens often, perhaps changing `@files`to be a `Hash(String,Int32)` would be a good idea? Your cache would probably still be worth it, but the rest of the method could be

```
if index = @files[name]?
  index
else
  @files[name] = @files.size 
end
```

.. but if it doesn't happen often then it would not be worth it.

Reply:
Yes, that's a good idea. I already tried it before sending this PR (we actually need a separate Hash because `@files` is used from the outside as an Array). It reduced the time from 0.82s to 0.79s, so maybe all the trouble of having more variables and more logic might not be worth it, but I can include it in this PR. What do you think?

Reply:
It's nice. Using a Hash *and* an Array will require more memory, but that should be contained since there aren't that much entries.

Reply:
Yes. Also, I had code that used the Hash just for computing this faster but then it's dropped because only the Array matters to clients (from what I've seen).

I'd really like to write test cases for all of these. I might do that these weeks.

Reply:
I added the hash.